File: 2015/rust/11.rs
// Advent of Code 2015, day 11
// (c) aichingert

fn update(cur: &mut Vec<char>) -> bool {
    let len = cur.len()-1;
    for i in 0..cur.len() {
        match cur[len-i] {
            'z' => cur[len-i] = 'a',
            _ => cur[len-i] = ((cur[len-i] as u8) + 1u8) as char,
        }

        if cur[len-i] != 'a' {
            break;
        }
    }

    if cur[0] == 'l' || cur[0] == 'o' || cur[0] == 'i' {
        return false;
    }

    let mut trip: bool = false;
    let mut amt: u8 = 0;
    let mut i = 0usize;

    for i in 0..cur.len()-2 {
        if (cur[i] as u8) + 1u8 == (cur[i+1] as u8) && (cur[i+1] as u8) + 1u8 == (cur[i+2] as u8) {
            trip = true;
            break;
        }
    }

    while i < cur.len()-1 {
        if cur[i+1] == 'l' || cur[i+1] == 'o' || cur[i+1] == 'i' {
            return false;
        }
        if cur[i] == cur[i+1] {
            amt += 1;
            i += 1;
        }
        i += 1;
    }

    trip && amt > 1
}

fn solve(cur: &mut Vec<char>) -> String {
    while !update(cur) {}
    cur.iter().map(|ch| *ch).collect::<String>()
}

fn main() {
	let inp = std::fs::read_to_string("../input/11").unwrap();
    let mut inp = inp.trim().chars().collect::<Vec<char>>();

    println!("Part 1: {}", solve(&mut inp));
    println!("Part 2: {}", solve(&mut inp));
}

File: 2015/julia/11.jl
# Advent of Code 2015, day 11
# (c) aichingert

module Day11

global pw = collect(readline("../input/11"))

function update()
	for i in eachindex(pw)
		if pw[length(pw)-(i-1)] == 'z' pw[length(pw)-(i-1)] = 'a'
		else pw[length(pw)-(i-1)] = Char(Int(pw[length(pw)-(i-1)]) + 1) end

		if pw[length(pw)-(i-1)] != 'a' break end
	end

	if pw[1] == 'l' || pw[1] == 'o' || pw[1] == 'i' return false end

	trip = false
	amt = 0

	for i in 1:length(pw)-2
		if Int(pw[i]) + 1 == Int(pw[i+1])  && Int(pw[i+1]) + 1 == Int(pw[i+2])
			trip = true
			break
		end
	end

	i = 1
	while i < length(pw)
		if pw[i+1] == 'l' || pw[i+1] == 'o' || pw[i+1] == 'i' return false end
		if pw[i] == pw[i+1]
			amt += 1
			i += 1
		end

		i += 1
	end
	
	trip && amt > 1
end

function solve()
	while !update() end
	String(pw)
end


println("Part 1: $(solve())")
println("Part 2: $(solve())")

end
